cmd/scaffold.go

```diff
@@ -837,27 +783,98 @@ func (fnb *FlowNodeBuilder) MustNot(err error) *zerolog.Event {
 	return nil
 }
 
-// Component adds a new component to the node that conforms to the ReadyDone
-// interface.
+// Component adds a new component to the node that conforms to the ReadyDoneAware
+// interface, and throws a Fatal() when an irrecoverable error is encountered
+// Use Component if the component cannot be restarted when an irrecoverable error is encountered
+// and the node should crash.
 //
 // When the node is run, this component will be started with `Ready`. When the
 // node is stopped, we will wait for the component to exit gracefully with
 // `Done`.
-func (fnb *FlowNodeBuilder) Component(name string, f func(builder NodeBuilder, node *NodeConfig) (module.ReadyDoneAware, error)) NodeBuilder {
-	fnb.components = append(fnb.components, namedComponentFunc{
-		fn:   f,
-		name: name,
+func (fnb *FlowNodeBuilder) Component(name string, f ComponentBuilderFunc) NodeBuilder {
+	fnb.componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		log := fnb.Logger.With().Str("component", name).Logger()
+
+		readyAware, err := f(fnb.NodeConfig)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("component %s initialization failed: %w", name, err))
+		}
+		log.Info().Msg("component initialization complete")
+
+		// TODO: switch to using RunComponent after we've refactored each of the ReadyDoneAware
+		//       objects to implement component.Component
+		if component, ok := readyAware.(component.Component); ok {
+			go component.Start(ctx)
+		}
+
+		if err := util.WaitClosed(ctx, readyAware.Ready()); err != nil {
+			log.Info().Msg("component startup aborted")
```

Comment:
I don't think it should crash, so that any components that did start successfully can be shutdown.

The old behavior was to return, which meant the `Done()` method was never called. In the new behavior, we continue and wait for the component to close it's done channel.

In theory it should be fine, but probably better to not call `Done()` for non-`Component`s